src/ZipkinBundle/Components/HttpClient/HttpClient.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+namespace ZipkinBundle\Components\HttpClient;
+
+use Throwable;
+use Zipkin\Tracer;
+use Zipkin\Tracing;
+use Zipkin\Propagation\Map;
+use const Zipkin\Tags\ERROR;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseStreamInterface;
+use Zipkin\SpanCustomizerShield;
+use ZipkinBundle\Components\HttpClient\ContextRetrievers\InScope;
+
+final class HttpClient implements HttpClientInterface
+{
+    /**
+     * @var HttpClientInterface
+     */
+    private $delegate;
+
+    /**
+     * @var Tracer
+     */
+    private $tracer;
+
+    /**
+     * @var callable
+     */
+    private $injector;
+
+    /**
+     * @var HttpParser
+     */
+    private $httpParser;
+
+    /**
+     * @var ContextRetriever
+     */
+    private $contextRetriever;
+
+    public function __construct(
+        HttpClientInterface $client,
+        Tracing $tracing,
+        ContextRetriever $contextRetriever = null,
+        HttpParser $httpParser = null
+    ) {
+        $this->delegate = $client;
+        $this->tracer = $tracing->getTracer();
+        $this->injector = $tracing->getPropagation()->getInjector(new Map());
+        $this->httpParser = $httpParser ?? new DefaultHttpParser();
+        $this->contextRetriever = $contextRetriever ?: new InScope($this->tracer);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        if ($this->contextRetriever === null) {
+            $span = $this->tracer->nextSpan();
+        } else {
+            $context = ($this->contextRetriever)($options);
+            $span = $this->tracer->nextSpan($context);
+        }
+
+        $spanCustomizer = new SpanCustomizerShield($span);
+        $this->httpParser->request($method, $url, $options, $spanCustomizer);
+
+        try {
+            $headers = [];
+            ($this->injector)($span->getContext(), $headers);
+            $options['headers'] = $headers + $options['headers'];
+            $response = $this->delegate->request($method, $url, $options);
+            $this->httpParser->response($response, $spanCustomizer);
```

Comment:
to simplify I usually parse response and error at the same time. add the thrown into the response and call parse when you have one or the other (or both)

what we do in brave now is defer adding the ERROR tag until we know it is zipkin being sent to. until then we keep a reference to the error.

Reply:
Makes sense.

Reply:
before, we did the union of error/response, but now we use custom response times in brave now, as it eases other things. you can decide what to do!

Reply:
In the end I can't have this signature because the response is async and we don't have access to it at this point.